pkg/tfshim/walk/walk.go

```diff
@@ -0,0 +1,198 @@
+	result := schema
+	for {
+		if len(p) == 0 {
+			return result, nil
+		}
+		nextResult, err := p[0].Lookup(result)
+		if err != nil {
+			return nil, fmt.Errorf("LookupSchemaPath failed at %s: %w", current, err)
+		}
+		result, p, current = nextResult, p[1:], current.WithStep(p[0])
+	}
+}
+
+// Similar to LookupSchemaPath but starts the initial step from a SchemaMap.
+func LookupSchemaMapPath(path SchemaPath, schemaMap shim.SchemaMap) (shim.Schema, error) {
+	return LookupSchemaPath(path, wrapSchemaMap(schemaMap))
+}
+
+// Represents elements of a SchemaPath.
+//
+// This interface is closed, the only the implementations given in the current package are allowed.
+type SchemaPathStep interface {
+	String() string
+	Lookup(shim.Schema) (shim.Schema, error)
+}
+
+// Drill down into an attribute by the given attribute name.
+type GetAttrStep struct {
+	Name string
+}
+
+func (step GetAttrStep) String() string {
+	return fmt.Sprintf("walk.GetAttrStep{%q}", step.Name)
+}
+
+func (step GetAttrStep) Lookup(s shim.Schema) (shim.Schema, error) {
+	if sm, ok := UnwrapSchemaMap(s); ok {
+		s, found := sm.GetOk(step.Name)
+		if !found {
+			return nil, fmt.Errorf("%s not found", step.String())
+		}
+		return s, nil
+	}
+	return nil, fmt.Errorf("%s is not applicable", step.String())
+}
+
+// Drill down into a Map, Set or List element schema.
+type ElementStep struct{}
+
+func (step ElementStep) String() string {
+	return "walk.ElementStep{}"
+}
+
+func (step ElementStep) Lookup(s shim.Schema) (shim.Schema, error) {
+	switch elem := s.Elem().(type) {
+	case shim.Resource:
+		switch s.Type() {
+		case shim.TypeMap:
+			return nil, fmt.Errorf("%s is not applicable to object types", step.String())
+		case shim.TypeList, shim.TypeSet:
+			return wrapSchemaMap(elem.Schema()), nil
+		default:
+			return nil, fmt.Errorf("%s is not applicable", step.String())
+		}
+	case shim.Schema:
+		return elem, nil
+	default:
+		return nil, fmt.Errorf("%s is not applicable", step.String())
+	}
+}
+
+func wrapSchemaMap(sm shim.SchemaMap) shim.Schema {
+	return (&schema.Schema{
+		Type: shim.TypeMap,
+		Elem: (&schema.Resource{Schema: sm}).Shim(),
+	}).Shim()
+}
+
+// Utility function to recognize nested object field type schemas encoded in shim.Schema.
+func UnwrapSchemaMap(s shim.Schema) (shim.SchemaMap, bool) {
+	switch elem := s.Elem().(type) {
+	case shim.Resource:
+		return elem.Schema(), true
+	default:
+		return nil, false
+	}
+}
+
+type SchemaVisitor = func(SchemaPath, shim.Schema)
+
+// Visit all nested schemas, including the current one.
+func VisitSchema(schema shim.Schema, visitor SchemaVisitor) {
+	walkSchemaInner(NewSchemaPath(), schema, visitor)
+}
+
+// Visit all nested schemas in a SchemaMap, keeping track of SchemaPath location.
+func VisitSchemaMap(schemaMap shim.SchemaMap, visitor SchemaVisitor) {
+	walkSchemaMapInner(NewSchemaPath(), schemaMap, visitor)
+}
+
+func walkSchemaInner(path SchemaPath, schema shim.Schema, visitor SchemaVisitor) {
```

Comment:
What is this `Inner` of? I don't see a `walkSchema`.
```suggestion
func walkSchema(path SchemaPath, schema shim.Schema, visitor SchemaVisitor) {
```